storage/cache/subtree_cache.go

```diff
@@ -151,23 +152,18 @@ func (s *SubtreeCache) stratumInfoForPrefixLength(l int) stratumInfo {
 func (s *SubtreeCache) preload(ids []storage.NodeID, getSubtrees GetSubtreesFunc) error {
 	// Figure out the set of subtrees we need.
 	want := make(map[string]storage.NodeID)
-	func() {
-		s.mutex.RLock()
-		defer s.mutex.RUnlock()
-
-		for _, id := range ids {
-			sInfo := s.stratumInfoForNodeID(id)
-			pxKey := id.PrefixAsKey(sInfo.prefixBytes)
-			// TODO(al): Fix for non-uniform strata.
-			id.PrefixLenBits = sInfo.prefixBytes * depthQuantum
-			if _, ok := s.subtrees[pxKey]; !ok {
-				want[pxKey] = id
-			}
+	for _, id := range ids {
+		sInfo := s.stratumInfoForNodeID(id)
+		pxKey := id.PrefixAsKey(sInfo.prefixBytes)
+		// TODO(al): Fix for non-uniform strata.
+		id.PrefixLenBits = sInfo.prefixBytes * depthQuantum
+		if _, ok := s.subtrees.Load(pxKey); !ok {
+			want[pxKey] = id
 		}
-	}()
-	// Note: At this point the lock is released, so multiple parallel preload
-	// invocations can happen to getSubtrees with the same node IDs. It's okay
-	// because we collapse results further below.
+	}
+	// Note: At this point multiple parallel preload invocations can happen to
+	// getSubtrees with overlapping sets of IDs. It's okay because we collapse
```

Comment:
It looks like multiple overlapping IDs will result in the same tree being fetched more than once?
It also looks like multiple overlapping IDs might invalidate the sync.map assumption that a particular key gets set once and then read many times.

Reply:
> It looks like multiple overlapping IDs will result in the same tree being fetched more than once?

Yes. This was also true before this PR. To completely avoid double-read of an ID we'd have to "lock" it while the database read is ongoing, which is quite a complex thing to implement without sacrificing performance (requires some finer grained locking than a "global" cache lock that we had previously). Instead, we can assume that this situation is rare. In fact, it doesn't seem to happen at all, if you consider how we use the cache: it's either read/written once (e.g. on different kinds of proof requests), or goroutines that write/read to/from it do not overlap w.r.t. subtrees (e.g. see the sparse Merkle tree building code).

> It also looks like multiple overlapping IDs might invalidate the sync.map assumption that a particular key gets set once and then read many times.

True. But I assume that this situation is rare or doesn't happen at all (see above). Obv., even if it does, `sync.Map` still works.